actors/builtin/miner/miner_test.go

```diff
@@ -782,58 +781,66 @@ func TestWindowPost(t *testing.T) {
 		advanceDeadline(rt, actor, &cronConfig{})
 	})
 
-	//runTillNextDeadline := func(rt *mock.Runtime) (*miner.DeadlineInfo, []*miner.SectorOnChainInfo, []uint64) {
-	//	st := getState(rt)
-	//	deadlines, err := st.LoadDeadlines(rt.AdtStore())
-	//	require.NoError(t, err)
-	//	deadline := actor.deadline(rt)
-	//
-	//	// advance to next deadline where we expect the first sectors to appear
-	//	rt.SetEpoch(deadline.NextOpen())
-	//	deadline = st.DeadlineInfo(rt.Epoch())
-	//
-	//	infos, partitions := actor.computePartitions(rt, deadlines, deadline.Index)
-	//	return deadline, infos, partitions
-	//}
+	t.Run("successful recoveries recover power", func(t *testing.T) {
+		rt := builder.Build(t)
 
-	//runTillFirstDeadline := func(rt *mock.Runtime) (*miner.DeadlineInfo, []*miner.SectorOnChainInfo, []uint64) {
-	//	actor.constructAndVerify(rt)
-	//
-	//	_ = actor.commitAndProveSectors(rt, 6, 100, nil)
-	//
-	//	// Skip to end of proving period, cron adds sectors to proving set.
-	//	actor.advancePastProvingPeriodWithCron(rt)
-	//
-	//	return runTillNextDeadline(rt)
-	//}
+		actor.constructAndVerify(rt)
+		infos := actor.commitAndProveSectors(rt, 1, 181, nil)
+		pwr := miner.PowerForSectors(actor.sectorSize, infos)
 
-	t.Run("successful recoveries recover power", func(t *testing.T) {
-		t.Skip("Disabled in miner state refactor #648, restore soon")
-		// TODO minerstate
-		//rt := builder.Build(t)
-		//deadline, infos, partitions := runTillFirstDeadline(rt)
-		//st := getState(rt)
+		// add lots of funds so we can pay penalties without going into debt
+		actor.addLockedFunds(rt, big.Mul(big.NewInt(200), big.NewInt(1e18)))
 
-		// mark all sectors as recovered faults
-		//sectors := bitfield.New()
-		//for _, info := range infos {
-		//	sectors.Set(uint64(info.SectorNumber))
-		//}
-		//err := st.AddFaults(rt.AdtStore(), &sectors, rt.Epoch())
-		//require.NoError(t, err)
-		//err = st.AddRecoveries(&sectors)
-		//require.NoError(t, err)
-		//rt.ReplaceState(st)
+		// Submit first PoSt to ensure we are sufficiently early to add a fault
+		// advance to next proving period
+		advanceAndSubmitPoSt(rt, actor, infos[0])
 
-		//pwr := miner.PowerForSectors(actor.sectorSize, infos)
-		//
-		//cfg := &poStConfig{
-		//	expectedRawPowerDelta: pwr.Raw,
-		//	expectedQAPowerDelta:  pwr.QA,
-		//	expectedPenalty:       big.Zero(),
-		//}
-		//
-		//actor.submitWindowPoSt(rt, deadline, partitions, infos, cfg)
+		// advance deadline and declare fault
+		advanceDeadline(rt, actor, &cronConfig{})
+		actor.declareFaults(rt, actor.declaredFaultPenalty(infos), infos...)
+
+		// advance a deadline and declare recovery
+		ongoingFee := actor.declaredFaultPenalty(infos)
+		advanceDeadline(rt, actor, &cronConfig{ongoingFaultsPenalty: ongoingFee})
```

Comment:
Thanks for this. I think you've highlighted an error in implementation.

When a sector is declared faulty and then recovered all in the span of a single proving period (so it misses zero PoSts), the total penalty should be just 1*FF. It looks like we're charging two. This one is the one that should be removed, with the recovery fee paid only on successful proof of the recovered sector.

As well as fixing, could you please make a final assertion of the actor's balance at the end of the test, so the total fee is clear?